src/repository.c

```diff
@@ -946,28 +954,48 @@ bool git_repository__reserved_names(
 }
 #endif
 
-static int check_repositoryformatversion(git_config *config)
+static int check_repositoryformatversion(int *version, git_config *config)
 {
-	int version, error;
+	int error;
 
-	error = git_config_get_int32(&version, config, "core.repositoryformatversion");
+	error = git_config_get_int32(version, config, "core.repositoryformatversion");
 	/* git ignores this if the config variable isn't there */
-	if (error == GIT_ENOTFOUND)
+	if (error == GIT_ENOTFOUND) {
+		giterr_clear();
 		return 0;
+	}
 
 	if (error < 0)
 		return -1;
 
-	if (GIT_REPO_VERSION < version) {
+	if (*version > GIT_REPO_VERSION_MAX_ALLOWED) {
 		giterr_set(GITERR_REPOSITORY,
 			"Unsupported repository version %d. Only versions up to %d are supported.",
-			version, GIT_REPO_VERSION);
-		return -1;
+			*version, GIT_REPO_VERSION_MAX_ALLOWED);
+		return GIT_EUNSUPPORTED;
 	}
 
 	return 0;
 }
 
+static int extension_each_cb(const git_config_entry *entry, void *payload)
+{
+	const char *name = entry->name + strlen("extension.");
+	(void)payload;
+
+	if (!strcmp(name, "noop"))
+		return 0;
+
+	giterr_set(GITERR_REPOSITORY, "Unknown Git repository extension: %s", name);
+	return GIT_EUNSUPPORTED;
+}
+
+static int check_extensions(git_repository *repo, git_config *config)
+{
+	return git_config_foreach_match(
+		config, "^extension\\..*", extension_each_cb, repo);
```

Comment:
The git-core spelling is `extensions` with an `s`. This needs fixed here, above in `extension_each_cb`, and in the tests.

I'm open to making it `extension` (I waffled between the two while writing the original), but we should make sure that git-core makes the same change. ;)